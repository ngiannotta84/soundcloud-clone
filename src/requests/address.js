// const address = "https://soundcloud-clone-api.onrender.com";
const address = "http://localhost:4000";

export default address;
