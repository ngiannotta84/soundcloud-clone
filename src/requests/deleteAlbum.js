import { deleteRequest } from "./helpers";

const deleteAlbum = async (id) => {
  const response = await deleteRequest("albums", id);
  return response;
};

export default deleteAlbum;
