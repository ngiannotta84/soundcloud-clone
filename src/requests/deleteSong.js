import { deleteRequest } from "./helpers";

const deleteSong = async (id) => {
  try {
    const response = await deleteRequest("songs", id);
    return response;
  } catch (err) {
    throw new Error(err);
  }
};

export default deleteSong;
