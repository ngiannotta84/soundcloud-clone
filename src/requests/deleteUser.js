import { deleteRequest } from "./helpers";

const deleteUser = async (id, password) => {
  const endpoint = `${id}/${password}`;

  const response = await deleteRequest("users", endpoint);
  return response;
};

export default deleteUser;
