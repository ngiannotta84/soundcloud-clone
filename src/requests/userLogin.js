import { postRequest } from "./helpers";

const userLogin = async (data) => {
  try {
    const response = await postRequest("users/login", data);
    return response;
  } catch (err) {
    throw new Error(err);
  }
};

export default userLogin;
