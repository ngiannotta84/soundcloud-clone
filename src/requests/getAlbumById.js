/* eslint-disable no-console */
import axios from "axios";
import address from "./address";

const getAlbumById = async (id) => {
  try {
    const response = await axios.get(`${address}/albums/${id}`);
    return response.data;
  } catch (err) {
    console.error(err);
    throw new Error(err);
  }
};

export default getAlbumById;
