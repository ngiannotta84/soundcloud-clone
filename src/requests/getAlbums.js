/* eslint-disable no-console */
import axios from "axios";
import address from "./address";

const getAlbums = async ({ name, exact, limit }) => {
  let endpoint = `${address}/albums`;
  const queryArray = [];
  if (name) {
    queryArray.push(`name=${name}`);
    if (exact) {
      queryArray.push("exact=true");
    }
  }
  if (limit) {
    queryArray.push(`limit=${limit}`);
  }
  if (queryArray.length > 0) {
    const query = queryArray.join("&");
    endpoint = `${endpoint}?${query}`;
  }
  console.log(endpoint);
  try {
    const response = await axios.get(endpoint);
    console.log(response);
    return response.data;
  } catch (err) {
    console.error(err);
    throw new Error(err);
  }
};

export default getAlbums;
