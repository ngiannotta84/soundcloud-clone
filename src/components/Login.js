import React, { useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import PropTypes from "prop-types";
import Alert from "./Alert";
import userLogin from "../requests/userLogin";
import "../styles/login.css";

const Login = ({ handleLogin }) => {
  const initialState = {
    fields: {
      email: "",
      password: "",
    },
    alert: "",
  };
  const [fields, setFields] = useState(initialState.fields);
  const [alert, setAlert] = useState(initialState.alert);
  const navigate = useNavigate();

  const handleFieldChange = (e) => {
    setFields({ ...fields, [e.target.name]: e.target.value });
  };

  const handleCredentials = async (event) => {
    const EMAIL_REGEX = /[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,4}$/;
    event.preventDefault();
    if (!fields.email) {
      setAlert("please provide your email address here");
    } else if (!fields.password) {
      setAlert("please insert your password here");
    } else if (!fields.email.match(EMAIL_REGEX)) {
      setAlert("please provide a valid email");
    } else {
      try {
        const response = await userLogin(fields);
        handleLogin(response);
        navigate(`/profile/${response.name}`);
      } catch (err) {
        setAlert(err.message);
      }
    }
  };

  return (
    <div className="login">
      <h2 className="please-login">Please Login</h2>
      <Alert message={alert} />
      <form onSubmit={handleCredentials}>
        <div className="form-fields">
          <div className="form-field">
            <label htmlFor="email">
              <h4 className="form-h4">email</h4>
              <input
                type="text"
                id="email"
                name="email"
                value={fields.email}
                onChange={handleFieldChange}
              />
            </label>
          </div>
          <div className="form-field">
            <label htmlFor="password">
              <h4 className="form-h4">password</h4>
              <input
                type="password"
                id="password"
                name="password"
                value={fields.password}
                onChange={handleFieldChange}
              />
            </label>
          </div>
        </div>
        <div>
          <button className="login-button" type="submit">
            Login
          </button>
        </div>
      </form>
      <p className="not-registered">
        Not Registered?
        <Link className="navbar-links-item-login" to="/signup">
          Sign-Up Here
        </Link>
      </p>
    </div>
  );
};

Login.propTypes = {
  handleLogin: PropTypes.func.isRequired,
};

export default Login;
