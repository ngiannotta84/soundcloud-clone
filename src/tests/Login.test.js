import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import Login from "../components/Login";
import * as userLogin from "../requests/userLogin";

describe("Login", () => {
  const validProps = {
    handleLogin: jest.fn(),
  };
  const userLoginMock = jest.spyOn(userLogin, "default");

  test("snapshot", () => {
    const { asFragment } = render(
      <Router>
        <Login handleLogin={validProps.handleLogin} />
      </Router>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe("tests", () => {
    beforeEach(() => {
      render(
        <Router>
          <Login handleLogin={validProps.handleLogin} />
        </Router>
      );
    });

    test("renders correctly", () => {
      expect(screen.getByText(/please login/i)).toBeInstanceOf(
        HTMLHeadingElement
      );
      expect(screen.getByLabelText(/email/i)).toBeInTheDocument();
      expect(screen.getByLabelText(/password/i)).toBeInTheDocument();
      expect(screen.getByText(/submit/i)).toBeInstanceOf(HTMLButtonElement);
      expect(screen.getByText(/not registered?/i)).toBeInTheDocument();
      expect(screen.getByText(/sign-up here/i)).toBeInTheDocument();
    });

    test("it renders an error message if no email is entered", () => {
      const button = screen.getByText("Submit");
      fireEvent.click(button);
      const message = screen.getByText(
        "please provide your email address here"
      );

      expect(message).toBeInTheDocument();
    });

    test("it renders an error message if no password is entered", () => {
      const button = screen.getByText("Submit");
      const email = screen.getByLabelText("email");

      fireEvent.change(email, { target: { value: "me@gmail.com" } });
      fireEvent.click(button);
      const message = screen.getByText("please insert your password here");

      expect(message).toBeInTheDocument();
    });

    test("it renders an error message if incorrect type of email is entered", () => {
      const button = screen.getByText("Submit");
      const email = screen.getByLabelText("email");
      const password = screen.getByLabelText("password");

      fireEvent.change(email, { target: { value: "megmail.com" } });
      fireEvent.change(password, { target: { value: "12345Nicola" } });
      fireEvent.click(button);
      const message = screen.getByText("please provide a valid email");

      expect(message).toBeInTheDocument();
    });

    test("it renders and error message if userLogin throws an error", () => {
      userLoginMock.mockImplementation(() => {
        throw new Error();
      });
      const button = screen.getByText("Submit");
      const email = screen.getByLabelText("email");
      const password = screen.getByLabelText("password");

      fireEvent.change(email, { target: { value: "valid@email.com" } });
      fireEvent.change(password, { target: { value: "validPassword" } });
      fireEvent.click(button);

      expect(
        screen.getByText(
          "Login failed, please recheck your email and password and try again"
        )
      ).toBeInTheDocument();
    });

    test("user is logged-in if userLogin is succesfull", async () => {
      const fakeResolve = {
        name: "fakeName",
        id: "fakeId",
      };
      userLoginMock.mockResolvedValue(fakeResolve);

      const button = screen.getByText("Submit");
      const email = screen.getByLabelText("email");
      const password = screen.getByLabelText("password");

      fireEvent.change(email, { target: { value: "valid@email.com" } });
      fireEvent.change(password, { target: { value: "validPassword" } });
      fireEvent.click(button);

      await waitFor(() => {
        expect(validProps.handleLogin).toBeCalledWith(fakeResolve);
      });
    });
  });
});
