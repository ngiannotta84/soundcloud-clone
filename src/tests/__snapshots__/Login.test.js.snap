// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login snapshot 1`] = `
<DocumentFragment>
  <div
    class="login"
  >
    <h2>
      Please Login
    </h2>
    <form>
      <div
        class="form-field"
      >
        <label
          for="email"
        >
          <span>
            email
          </span>
          <input
            id="email"
            name="email"
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        class="form-field"
      >
        <label
          for="password"
        >
          <span>
            password
          </span>
          <input
            id="password"
            name="password"
            type="password"
            value=""
          />
        </label>
      </div>
      <div>
        <button
          class="submit-button"
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
    <p>
      Not Registered?
      <a
        class="navbar-links-item"
        href="/signup"
      >
        Sign-Up Here
      </a>
    </p>
  </div>
</DocumentFragment>
`;
